li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -565,19 +570,79 @@ public void uncaughtException(Thread t, Throwable e) {
   }
 
-  synchronized public void reloadConfig(Map<String, Object> newConfigs) {
+  public LiKafkaConsumerConfig getCommonConsumerConfigs() {
+    return _commonConsumerConfigs;
+  }
+
+  public void reloadConfig(Map<String, String> newConfigs, UUID commandId) {
+    // Go over each consumer, close them, and update existing configs with newConfigs. Since each per-cluster consumer will be
+    // instantiated when the client began consuming from that cluster, we just need to clear the mappings and update the configs.
+    // since this method is invoked by marioClient, it should be non-blocking, so we create another thread doing above
+    LOG.info("Starting reloadConfig for LiKafkaConsumers in clusterGroup {} with new configs {}", _clusterGroup, newConfigs);
+    Thread t = new Thread(() -> {
+      closeExistingConsumers(newConfigs, commandId);
+    });
+    t.setDaemon(true);
+    t.setName("LiKafkaConsumer-reloadConfig-" + _clusterGroup.getName());
+    t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+      public void uncaughtException(Thread t, Throwable e) {
+        throw new KafkaException("Thread " + t.getName() + " throws exception", e);
+      }
+    });
+
+    t.start();
+  }
+
+  // Do synchronization on the method that will be closing the consumers since it will be handled by a different thread
+  synchronized void closeExistingConsumers(Map<String, String> newConfigs, UUID commandId) {
     ensureOpen();
-    // TODO: Go over each consumer, close, and recreate them
+
+    closeNoLock(RELOAD_CONFIG_EXECUTION_TIME_OUT.toMillis(), TimeUnit.MILLISECONDS);
+
+    // update existing configs with newConfigs
+    // originals() would return a copy of the internal consumer configs, put in new configs and update existing configs
+    Map<String, Object> configMap = _commonConsumerConfigs.originals();
+    configMap.putAll(newConfigs);
+    _commonConsumerConfigs = new LiKafkaConsumerConfig(configMap);
+
+    // reset the state of _consumers and _closed so that subsequent calls will try to instantiate the consumers with new configs
+    _consumers.clear();
+    _closed = false;
+
+    // Convert the updated configs from Map<String, Object> to Map<String, String> and send the new config to mario server
+    // report reload config execution complete to mario server
+    Map<String, String> convertedConfig = new HashMap<>();
+    for (Map.Entry<String, Object> entry : configMap.entrySet()) {
+      convertedConfig.put(entry.getKey(), String.valueOf(entry.getValue()));
+    }
+    _mdsClient.reportCommandExecutionComplete(commandId, convertedConfig, MsgType.RELOAD_CONFIG_RESPONSE);
+
+    // re-register federated client with updated configs
+    _mdsClient.reRegisterFederatedClient(newConfigs);
+
+    _latchForTest.countDown();
+
+    LOG.info("Successfully restarted LiKafkaConsumers in clusterGroup {} with new configs (diff) {}", _clusterGroup, newConfigs);
```

Comment:
Nit: to be exact, this didn't restarted consumers. Maybe we should say successfully updated (or reloaded) configs?